pkg/util/api/api.go

```diff
@@ -38,3 +42,32 @@ func truncateMessage(message string, limit int) string {
 	suffix := " ..."
 	return message[:limit-len(suffix)] + suffix
 }
+
+// SetContainersDefaults will fill up the container defaults.
+// Note that this is used to bridge the gap between job and workload,
+// E.g. job may have resources limits configured but no requests,
+// this is not allowed in workload but we can't update the job directly,
+// so we need a function like this to set the defaults.
+func SetContainersDefaults(containers []v1.Container) []v1.Container {
+	// This should only happen in tests.
+	// We didn't return nil here for spec.Containers should not be nil,
+	// we also didn't return empty slice directly here for some unit tests expect this to be nil.
+	if len(containers) == 0 {
+		return containers
+	}
+	res := make([]v1.Container, len(containers))
+	for i, c := range containers {
```

Comment:
Let's add a comment that this logic is copied from k8s defaulting https://github.com/kubernetes/kubernetes/blob/ab002db78835a94bd19ce4aaa46fb39b4d9c276f/pkg/apis/core/v1/defaults.go#L144